New Checkout: split purchase details into second column

michaeldcain opened the pull request:
This does a few things with the ultimate goal of splitting out a summary into a separate column on desktop screens.

- adds the floated `CheckoutSummary` component to composite-checkout (package, demo, documentation)
- moves `CheckoutStepBody` component inside `CheckoutSteps`
- uses `CheckoutSummary` in WPcom's implementation of composite-checkout
- moves the current site's address to the review step
- adds a default set of purchase features to the summary (per-product features coming in a future PR)
- hides `CheckoutSummary` on mobile screens (collapse mobile toggle coming in future PR)

Before | After
------------ | -------------
<img width="1164" alt="before-desktop" src="https://user-images.githubusercontent.com/942359/79905306-05b8ca00-83e4-11ea-9c19-7f66901e6e2e.png"> | <img width="1164" alt="after-desktop" src="https://user-images.githubusercontent.com/942359/79905305-05203380-83e4-11ea-8692-4fd43aee5769.png">
<img width="374" alt="before-mobile" src="https://user-images.githubusercontent.com/942359/79905302-05203380-83e4-11ea-9225-1bca7db43ff0.png"> | <img width="372" alt="after-mobile" src="https://user-images.githubusercontent.com/942359/79905300-04879d00-83e4-11ea-936b-54c262a6ae4e.png">

Fixes: #41088

**To test:**
- it might help to step through the PR, commit by commit, but no promises that I didn't change my mind later on in the PR
- run the composite-checkout demo with `yarn run composite-checkout-demo`
- verify that minimal CheckoutSummary is displayed
- verify that the demo otherwise works as expected
- from dev or calypso.live, and a plan and/or domain to your cart and visit Checkout with `?flags=composite-checkout-testing`
- verify that on desktop (`>960px`), the checkout summary is displayed, floated to the right, with a list of default features, and the tax (once calculated) and total
- verify that on screens smaller than desktop, the checkout summary is hidden
- verify that Checkout otherwise works as expected

michaeldcain committed:
Move domain to review step pull purchase deats out of checkout wrapper

michaeldcain committed:
Add useTax

michaeldcain committed:
Add tax and basic styling to order summary

michaeldcain committed:
Adjust summary amount styles

sirbrillig commented on client/my-sites/checkout/composite-checkout/wpcom/components/wp-order-review-line-items.js:
```diff
@@ -310,23 +310,14 @@ WPOrderReviewLineItems.propTypes = {
 const WPOrderReviewList = styled.ul`
-	margin: -10px 0 10px 0;
-	padding: 0;
+	margin: 10px 0;
```
What is the purpose of these CSS changes? It's not mentioned in the commit, and it seems pretty significant since these styles are used for the line items in the review step.

michaeldcain replied:
Sorry, the `-10px` top margin was used to visually make the list items "nest" under the step title. In the new design, it's been [replaced by the `domainUrl`](https://github.com/Automattic/wp-calypso/pull/41063/files#diff-8503f114a52569d2dee502c3a31acb6bR74).

sirbrillig commented on client/my-sites/checkout/composite-checkout/wpcom/components/wp-checkout-order-review.js:
```diff
@@ -12,6 +12,7 @@ import { useLineItems, useFormStatus } from '@automattic/composite-checkout';
 import joinClasses from './join-classes';
 import Coupon from './coupon';
 import { WPOrderReviewLineItems, WPOrderReviewSection } from './wp-order-review-line-items';
+import { isLineItemADomain } from '../hooks/has-domains';
```
Pretty minor but ideally the title of this commit shouldn't start with a `-`.

sirbrillig commented on packages/composite-checkout/src/lib/line-items.js:
```diff
@@ -25,3 +25,8 @@ export function useTotal() {
 	return total;
 }
+
+export function useTax() {
```
Good idea! I wonder, though, if this should just be a generalized hook like `useLineItemsOfType('tax')` and/or `useFirstLineItemOfType('tax')` (there may be better names) since `composite-checkout` has no concept of "tax" types (or any type at all, really). Types are currently designed to be for the use of the consumer.

michaeldcain committed:
Remove WPCheckoutOrderSummaryTitle for now.

michaeldcain committed:
Missing propType

michaeldcain committed:
Switch useTax to useFirstLineItemOfType hook

michaeldcain committed:
Remove unnecessary span

michaeldcain committed:
Return all line items of a certain type.

michaeldcain committed:
Move MainContentUI to CheckoutSteps component

michaeldcain committed:
Restructure Checkout and CheckoutSteps to allow for CheckoutSummary

- Move CheckoutStepBody inside CheckoutSteps
- Split real steps from not
- Move `checkout-content` back to Checkout component
- Update loading
- Update demo

michaeldcain committed:
Add CheckoutSummary component skeleton with defaults and demo

michaeldcain committed:
Cleanup CheckoutSummary

- Remove props from Checkout Summary so that component just renders children
- Update demo and defaults
- Fix summary height when split into columns

michaeldcain committed:
Cleanup styles up to this point

michaeldcain committed:
Make CheckoutSummary default content unique from CheckoutSummaryStep

michaeldcain committed:
Remove unused styled components from default summary

michaeldcain committed:
Use CheckoutSummary in WPcom Checkout

michaeldcain committed:
Remove widths from MainContentUI to allow for optional CheckoutSummary

michaeldcain committed:
Update documentation to match current state

michaeldcain committed:
Update tests to reflect order summary step > checkout summary change

michaeldcain committed:
Center checkout for single-column layout

michaeldcain committed:
Tighten up order review line item styles to match design

michaeldcain committed:
Export CheckIcon and allow it to be styled

michaeldcain committed:
Stub out CheckoutSummaryFeatures for initial styling

a8ci18n commented on client/my-sites/checkout/composite-checkout/wpcom/components/wp-checkout-order-summary.js:
```diff
@@ -3,46 +3,109 @@
+				<CheckoutSummaryFeaturesList>
+					<CheckoutSummaryFeaturesListItem>
+						<WPCheckoutCheckIcon />
+						{ translate( 'Live chat and email support' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **29** times:
`translate( 'Email and live chat support' )` *ES Score: 9*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/checkout-two-column-layout)

a8ci18n commented on client/my-sites/checkout/composite-checkout/wpcom/components/wp-checkout-order-summary.js:
```diff
@@ -3,46 +3,109 @@
+					</CheckoutSummaryFeaturesListItem>
+					<CheckoutSummaryFeaturesListItem>
+						<WPCheckoutCheckIcon />
+						{ translate( 'Dozens of free themes' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **19** times:
`translate( 'Dozens of free themes.' )` *ES Score: 15*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/checkout-two-column-layout)

a8ci18n commented on client/my-sites/checkout/composite-checkout/wpcom/components/wp-checkout-order-summary.js:
```diff
@@ -3,46 +3,109 @@
+					</CheckoutSummaryFeaturesListItem>
+					<CheckoutSummaryFeaturesListItem>
+						<WPCheckoutCheckIcon />
+						{ translate( 'Money back guarantee' ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **17** times:
`translate( '30-day Money Back Guarantee' )` *ES Score: 9*

michaeldcain committed:
Hide summary on mobile

sirbrillig commented on packages/composite-checkout/README.md:
```diff
@@ -315,6 +342,10 @@ A React Hook that will return true if the current step is complete as defined by
 
+### useLineItemsOfType
+
+A React Hook that will return an array of line items matching a specific 'type' (i.e. 'tax'). Only works within [CheckoutProvider](#CheckoutProvider).
```
We should probably explain that it takes a string argument and what the `type` means. I'd say something like:

```md
A React Hook taking one string argument that will return an array of [line items](#line-items) from the cart (derived from the same data returned by [useLineItems](#useLineItems)) whose `type` property matches that string. Only works within [CheckoutProvider](#CheckoutProvider).
```

sirbrillig commented on client/my-sites/checkout/composite-checkout/wpcom/components/wp-checkout-order-review.js:
```diff
@@ -23,13 +25,20 @@ export default function WPCheckoutOrderReview( {
+
 	return (
 		<div className={ joinClasses( [ className, 'checkout-review-order' ] ) }>
+			{ domainUrl && <DomainURL>{ translate( 'Site' ) + ': ' + domainUrl }</DomainURL> }
```
I have some concerns about how other locales (particularly RTL languages) might handle this translation (we should probably include the url as a replacement string in the translation). Even though we could probably wait until the i18n phase, it's probably better to do this now. See [here](https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/checkout/composite-checkout/wpcom/components/wp-order-review-line-items.js#L229) for an example.

sirbrillig commented on packages/composite-checkout/src/components/checkout-steps.js:
```diff
@@ -153,25 +157,41 @@ export function CheckoutSteps( { children } ) {
+				<CheckoutErrorBoundary
+					errorMessage={ localize( 'There was a problem with the submit button.' ) }
+				>
+					<CheckoutSubmitButton disabled={ isThereAnotherNumberedStep || formStatus !== 'ready' } />
```
So because we're moving the submit button into `CheckoutSteps`, that makes `CheckoutSteps` pretty much required for the form. We should note that early in the README somehow; both that it creates the submit button previously included by `Checkout` and that if you don't use it you won't get a submit button. (That actually might not be clear in the current version of the README but I think it's more important now since we have so many "helpful components you can use" which really are not so optional.) Effectively, `CheckoutSteps` has become the form now rather than "a group of steps joined by continue buttons" so we might want to update its description as well.

sirbrillig commented on packages/composite-checkout/src/lib/line-items.js:
```diff
@@ -25,3 +25,11 @@ export function useTotal() {
 	return total;
 }
+
+export function useLineItemsOfType( itemType = null ) {
```
I don't think we need to provide a default value of `null` if it's actually required (JS won't complain about missing arguments anyway).

sirbrillig reviewed (commented):
The code is looking good! Made some small suggestions. Tested both the demo and the calypso version and both look good to me!

michaeldcain committed:
Alphabetize

michaeldcain committed:
Improve docs for useLineItemsOfType

michaeldcain committed:
Provide domainUrl as replacement string in translation

michaeldcain committed:
Remove default value on useLineItemsOfType hook

michaeldcain committed:
Update docs to reflect moving the form structure to CheckoutSteps

michaeldcain commented:
@sirbrillig: thanks for the review. I believe I've addressed your feedback and this is ready for another go.

sirbrillig reviewed (approved):
Works as described! Thanks!

manzoorwanijk commented:
A false positive here. I linked a Jetpack issue to #98930 which as the same issue ID as this PR 😄
